### ISSUE TITLE
Progress messages system overhaul

### ISSUE THREAD

**Issue by bertfrees** [daisy/pipeline-tasks#102]
There are a couple of things about the current implementation of progress indication (approach explained in [daisy/pipeline#445](https://github.com/daisy/pipeline/issues/445#issuecomment-207341862)) that I'm not completely happy with. I will list them here and explain how I plan to improve things.

## Log level vs "progress depth"

With "progress depth" I mean the level of nested "steps", where a "step" is basically a named segment of the progress bar. The depth has two purposes:

- to model the progress
- to present messages in a structured way to the user: the depth is translated to an indentation level in the UI

In addition to this, we are also using depth as a way to hide (with a expand/collapse button) deeply nested messages because the depth can be an indication of the importance of messages. For example, a use case is a script that uses another script internally to convert an intermediary format, and the user is not interested in all the details of that conversion. By collapsing messages deeper than e.g. level 2 or 3, the details of the inner script are automatically hidden, while if the script was used by itself they would not be hidden.

The problem with this is that the depth of messages doesn't necessarily determine the importance. We could be hiding too much or too little. Moreover, sometimes it is simply not desired to show the complete structure to the user (even if it's collapsed by default). The importance of messages should be determined mainly by the log level, and the log level of messages should be the responsibility of the script.

I want to solve this as follows:

- Make it possible for an XProc step, when invoking another step, to set a threshold on the level of messages from the inner step to come through. The suppressed messages are not discarded but changed to TRACE level so that they can still be enabled by the user if needed.
- Make it possible for an XProc step, when invoking another step, to control whether or not the inner structure is presented to the user, because in some cases this is not relevant and even undesired. In other words, the possibility to "flatten" the structure, or "promote" inner messages. When promoting a message it should be possible to add a prefix. The original structure may still be traced in the log file. This is kind of similar to how exception can be caught and new exceptions can be created from them.

See below (Explicitness) for the practical solution.

## API

Currently the progress information is encoded as a special messages inside square brackets (`[progress ...]`). This is kind of a workaround. Much nicer would be if this info would be a full part of the API. In the web API, this means additional attributes on the `message` element. In the Java API, this means additional fields in the Message objects.

An additional benefit of this is that clients won't have to actively remove progress info from messages like is currently the case. Progress info will be ignored by default. Clients will only need to do special processing if they want to implement progress indication.

How progress info is represented internally is a different issue. It makes the most sense to represent progress messages as special purpose objects in Java and XML in XProc. Simple messages can still be represented as log messages (string and level) in order to make full use of SLF4J. In addition we could possibly still support the special `[progress ...]` encoded messages as well (but only internally).

## Explicitness

The main drawback of the current approach for specifying progress info in scripts is that it is too explicit:

- It hinders readability.
- Naming the "progress steps" is manual work, an additional thing to think about, and a source of error.
- It is confusing because the new concept of "progress steps" lies very close to XProc steps/groups but is not exactly the same.
- There is the issue of XProc pipelines being suitable for parallelization so that XProc engines (Calabash) may execute steps of a pipeline in an unpredictable way. For this reason, the actual execution of a "progress step" and its corresponding message may not happen directly after each other in time. This is where the "cx:depends-on" workaround in some places came from.

My solution is to make everything more natural and implicit by replacing the `px:message` calls (and the `px:progress` utility step we had planned) with custom attributes on XProc step invocations (including `p:identity`) and on things like `p:group`. The explicit naming and linking is not required anymore because names can automatically be inferred from the `name` attributes or otherwise generated, and passed along to children. Implicit progress steps can be generated around each iteration of a `p:for-each` with an increment of `1 / p:iteration-size()`. This will all be done by the XProc engine.

As opposed to the old system where progress steps were automatically closed when the next steps starts, the new system requires that they are closed explicitly (explicit in terms of progress info sent to the client, but it would be done automatically by the engine). The old way could break if the order of execution in a pipeline is shuffled by the engine.

The custom attributes are:

- `px:progress`: progress increment as a fraction (`x/x`) or percentage (`x%`)

- `px:message`: in order to connect a message with progress info

- `px:message-severity`: like `severity` attribute on `px:message` step

- `px:short-message`: string to prepend to inner messages when they are promoted

  Messages are promoted when no `px:message` is specified, or when the message is not rendered because its level (see `px:message-severity`) falls below the threshold which is determined by the user configuration (logback.xml) and the `px:filter-messages` values of enclosing steps (see below). A non-empty prefix is separated from the message with a colon. Defaults to "".

- `px:filter-messages`: change the severity level of inner messages to "TRACE" if below this threshold

  This includes also inner messages of inner steps. The original level and a colon are prepended to the message when made into a TRACE message (unless the original level was TRACE). Defaults to "ALL" (no messages filtered out).

The main shortcoming of this new approach is that attributes are less powerful than options because they can not be provided dynamically. For `px:progress` this should normally be no issue because the progress does normally not need to be computed dynamically. For the `px:message` attribute this could possibly be an issue. A solution could be to put a simple message on the step and if needed a more specific one inside the step. Alternatively, we could look into supporting attribute value templates (curly braces).

**Comment by josteinaj** [comment]
Sounds great. Some comments (no objections):

*API*: One consideration would be how the "current progress" is calculated. I think we have timestamps for each message in the Web API now. I previously had that implemented in the Web UI as a workaround for timestamps not being present in the web api. The timestamps are used to estimate how fast the progress is moving, so that you know where between two steps you are. The most basic approach would be to still leave this interpolation up to the client. Another approach would be to have a global "current progress" for a job, where the interpolation happens in the engine. "current progress" could possibly be a separate endpoint, so that the interpolation doesn't happen unless you want it to (i.e. `/jobs/{id}/progress`).

*Explicitness*: Note that `cx:message` is now also available as an attribute in vanilla Calabash, which makes the unpredictable execution of steps less of a problem (https://github.com/ndw/xmlcalabash1/issues/238). It is available for us now that you've upgraded to a newer version of Calabash. But the other attributes you describe still sound very useful in addition.

*Shortcoming regarding dynamically provided messages*: I think this is fine. We rarely (maybe never) need this. Maybe there could be a workaround using the px:message step in such cases, but I wouldn't worry about it. I believe the only places where dynamic messages are used now are for places where `1 / p:iteration-size()` is inserted, and in the braille scripts where the substep name `px:transform` is assumed based on the transform query. But those things wouldn't need to be computed dynamically with the approach as you describe it.

**Comment by bertfrees** [comment]
> Another approach would be to have a global "current progress" for a job, where the interpolation happens in the engine.

Yes, I have thought about that too. It could be nice. Would make the clients lighter for sure.

My only concern was that a single progress value is maybe not rich enough for clients that want more then a simple linear progress bar, but maybe that is a small drawback for what we gain.

Another issue is that if you don't have the "progress" info on messages anymore, it also doesn't really make sense to have the "progress-of" and "progress-within" values which are used by the client to get the structure of messages. So I thought about simply sending a "depth" value instead (which is currently also computed in the client). However this is not sufficient for more complex structures, notably when there are multiple branches progressing in parallel, which could happen in theory (not in practice I think).

How should we handle this case? It is pretty hard for clients to render this kind of structure anyway, so maybe we should just assert somehow that this never happens?

Alternatively we could have a "parent" attribute which would point to the sequence number of the parent message. ("sequence" and "parent" would basically be the same as "progress-of" and "progress-within" except named differently and their values would be numbers, not just any string.)

> Note that `cx:message` is now also available as an attribute in vanilla Calabash.

Ah thanks, didn't realize that. I'll check if we can reuse it.

**Comment by josteinaj** [comment]
I meant that the global progress could be *in addition* to whatever else is added to the job XML. It would then be up to the client if it wants to use that, or something more advanced.

The message/progress tree could also be created by the engine. Either replace the flat message structure we have now or create a new endpoint:
```
<messages>
    <message sequence="1" content="Starting script" .../>
    <message sequence="2" content="Starting parallel process 1/2" ...>
        <message sequence="4" content="whatever..." ...>
            <message sequence="5" content="whatever..." .../>
        </message>
        <message sequence="7" content="whatever..." .../>
    </message>
    <message sequence="3" content="Starting parallel process 1/2" ...>
        <message sequence="6" content="whatever..." .../>
        <message sequence="8" content="whatever..." .../>
    </message>
</messages>
```

The client could then render by sequence (i.e. time), or as a nested tree structure.

**Comment by bertfrees** [comment]
I thought about that too :)

What should happen when messages starting from a certain sequence number are requested?

**Comment by josteinaj** [comment]
Hmm. I guess "ancestor messages" that would normally not be included would have to be included. i.e. `ancestor::message`, but `preceding::message` could still be excluded.

**Comment by bertfrees** [comment]
Yeah, that could work.

> https://github.com/ndw/xmlcalabash1/issues/238

By the way I had completely forgotten we discussed this solution before! But it must have stuck in the back of my mind.

In his fix Norman simply prints to stderr though, so I'll have to extend it.

**Comment by bertfrees** [comment]
Note that this will make the API backwards incompatible.

**Comment by bertfrees** [comment]
Possibly I could make Calabash make debug messages from `p:documentation` defined inside step invocations, groups, etc.

**Comment by josteinaj** [comment]
Regarding backwards incompatibility: the nested stucture could be enabled in the url with `&nested=1` or similar, thus the default would be a backwards compatibile flat structure.

**Comment by bertfrees** [comment]
OK.

**Comment by bertfrees** [comment]
I think I have figured out the API. When `&nested=0` (default), the client will get a flat and linear structure that is backwards-compatible (if clients ignore the `progress` attribute). Messages that would normally be nested are promoted.

```xml
<messages progress="0.42">
  <message sequence="1">Starting script</message>
  <message sequence="2">Starting parallel process 1/2</message>
  <message sequence="3">Starting parallel process 2/2</message>
  <message sequence="4">1/2: whatever...</message>
  <message sequence="5">1/2: whatever...</message>
  <message sequence="6">2/2: whatever...</message>
  <message sequence="7">1/2: whatever...</message>
  <message sequence="8">2/2: whatever...</message>
</messages>
```

When `&nested=1`, the client will get a nested structure like so:

```xml
<messages progress="0.42">
  <message portion="0.1" progress="1" sequence="1" content="Starting script"/>
  <message portion="0.2" progress="0.55" sequence="2" content="Starting parallel process 1/2">
    <message portion="0.5" progress="1" sequence="4" content="whatever...">
      <message sequence="5" content="whatever..."/>
    </message>
    <message portion="0.1" progress="0.5" sequence="7" content="whatever..."/>
  </message>
  <message portion="0.1" progress="0.8" sequence="3" content="Starting parallel process 2/2">
    <message portion="0.4" progress="1" sequence="6" content="whatever..."/>
    <message portion="0.4" progress="1" sequence="8" content="whatever..."/>
  </message>
</messages>
```

`portion` is relative to the parent `portion`, `progress` is relative to the own `portion`. The `progress` of a message is always greater than or equal to the sum of `progress` times `portion` over its child messages (the difference is the progress not associated with any messages):

∑<sub>children</sub> (`portion` × `progress`) ≤ `progress` ≤ 1 - ∑<sub>children</sub> (`portion` × (1 - `progress`))

`portion` of the top-level `messages` element is implied ("1"). For `message` elements, an absent `portion` means "0" (and `progress` is absent too).

**Comment by bertfrees** [comment]
Interpolation I would for now leave up to the client because it doesn't really make the progress info any more precise. I think it is purely a visual/UI thing to give the user the feeling that something is happening and to make the progress more fluent. But how fluent it will make the progress still depends on how often the client checks new messages.

We can still decide to add it to the server later.

Initially I'm even thinking about dropping all the new attributes except the top-level `progress` to keep things simple. The web UI currently would only use that information to do interpolation, and it might do fine without.

**Comment by bertfrees** [comment]
@rdeltour FYI
